cli/src/lib.rs

```diff
@@ -530,56 +483,250 @@ fn genesis_domain(public_key: PublicKey) -> Domain {
-/// - if config fails to build
-pub fn combine_configs(args: &Arguments) -> color_eyre::eyre::Result<Configuration> {
-    args.config_path
-        .first_existing_path()
-        .map_or_else(
-            || {
-                eprintln!("Configuration file not found. Using environment variables as fallback.");
-                ConfigurationProxy::default()
-            },
-            |path| {
-                let path_proxy = ConfigurationProxy::from_path(&path.as_path());
-                // Override the default to ensure that the variables
-                // not specified in the config file don't have to be
-                // explicitly specified in the env.
-                ConfigurationProxy::default().override_with(path_proxy)
-            },
-        )
-        .override_with(
-            ConfigurationProxy::from_std_env()
-                .wrap_err("Failed to build configuration from env")?,
-        )
+/// - If config fails to build
+/// - If genesis config is invalid
+pub fn read_config(
+    path: &Path,
+    submit_genesis: bool,
+) -> Result<(Configuration, Option<GenesisNetwork>)> {
+    let config = ConfigurationProxy::default();
+
+    let config = if let Some(actual_config_path) = path
+        .try_resolve()
+        .wrap_err("Failed to resolve configuration file")?
+    {
+        let mut cfg = config.override_with(ConfigurationProxy::from_path(&*actual_config_path));
+        let config_dir = actual_config_path
+            .parent()
+            .expect("If config file was read, than it should has a parent. It is a bug.");
+
+        // careful here: `genesis.file` might be a path relative to the config file.
+        // we need to resolve it before proceeding
+        // TODO: move this logic into `iroha_config`
+        //       https://github.com/hyperledger/iroha/issues/4161
+        let join_to_config_dir = |x: &mut PathBuf| {
+            *x = config_dir.join(&x);
+        };
+        mutate_nested_option!(cfg, [genesis, file, self], join_to_config_dir);
+        mutate_nested_option!(cfg, [snapshot, dir_path], join_to_config_dir);
+        mutate_nested_option!(cfg, [kura, block_store_path], join_to_config_dir);
+        mutate_nested_option!(cfg, [telemetry, file, self], join_to_config_dir);
+
+        cfg
+    } else {
+        config
+    };
+
+    // it is not chained to the previous expressions so that config proxy from env is evaluated
+    // after reading a file
+    let config = config.override_with(
+        ConfigurationProxy::from_std_env().wrap_err("Failed to build configuration from env")?,
+    );
+
+    let config = config
         .build()
-        .map_err(Into::into)
+        .wrap_err("Failed to finalize configuration")?;
+
+    // TODO: move validation logic below to `iroha_config`
+
+    if !submit_genesis && config.sumeragi.trusted_peers.peers.len() < 2 {
+        return Err(eyre!("\
+            The network consists from this one peer only (`sumeragi.trusted_peers` is less than 2). \
```

Comment:
The `fails_with_no_trusted_peers_and_submit_role` test fails because of the updated test message

Reply:
Thanks, fixed.